webknossos-datastore/app/com/scalableminds/webknossos/datastore/helpers/DatasetDeleter.scala

```diff
@@ -31,23 +39,244 @@ trait DatasetDeleter extends LazyLogging with DirectoryConstants {
         case e: Exception                                => Fox.failure(s"Deleting dataset failed: ${e.toString}", Full(e))
       }
 
+    def moveToTrash(organizationId: String,
+                    datasetName: String,
+                    dataSourcePath: Path,
+                    reason: Option[String]): Fox[Unit] =
+      if (Files.exists(dataSourcePath)) {
+        val trashPath: Path = dataBaseDir.resolve(organizationId).resolve(trashDir)
+        val targetPath = trashPath.resolve(datasetName)
+        new File(trashPath.toString).mkdirs()
+
+        logger.info(s"Deleting dataset by moving it from $dataSourcePath to $targetPath${if (reason.isDefined)
+          s" because ${reason.getOrElse("")}"
+        else "..."}")
+        deleteWithRetry(dataSourcePath, targetPath)
+      } else {
+        Fox.successful(logger.info(
+          s"Dataset deletion requested for dataset at $dataSourcePath, but it does not exist. Skipping deletion on disk."))
+      }
+
     val dataSourcePath =
       if (isInConversion) dataBaseDir.resolve(organizationId).resolve(forConversionDir).resolve(datasetName)
       else dataBaseDir.resolve(organizationId).resolve(datasetName)
 
-    if (Files.exists(dataSourcePath)) {
-      val trashPath: Path = dataBaseDir.resolve(organizationId).resolve(trashDir)
-      val targetPath = trashPath.resolve(datasetName)
-      new File(trashPath.toString).mkdirs()
+    for {
+      _ <- moveSymlinks(organizationId, datasetName) ?~> "Failed to remake symlinks"
+      _ <- moveToTrash(organizationId, datasetName, dataSourcePath, reason)
+    } yield ()
+  }
 
-      logger.info(
-        s"Deleting dataset by moving it from $dataSourcePath to $targetPath${if (reason.isDefined) s" because ${reason.getOrElse("")}"
-        else "..."}")
-      deleteWithRetry(dataSourcePath, targetPath)
+  def remoteWKClient: Option[DSRemoteWebknossosClient]
+
+  // Handle references to layers and mags that are deleted
+
+  private def moveSymlinks(organizationId: String, datasetName: String)(implicit ec: ExecutionContext) =
+    for {
+      dataSourceId <- Fox.successful(DataSourceId(datasetName, organizationId))
+      layersAndLinkedMagsOpt <- Fox.runOptional(remoteWKClient)(_.fetchPaths(dataSourceId))
+      exceptionBoxes = layersAndLinkedMagsOpt match {
+        case Some(layersAndLinkedMags) =>
+          layersAndLinkedMags.map(layerMagLinkInfo =>
+            handleLayerSymlinks(dataSourceId, layerMagLinkInfo.layerName, layerMagLinkInfo.magLinkInfos.toList))
+        case None => Seq(tryo {})
```

Comment:
what does the `tryo {}` return? maybe we could make this more explicit?

Reply:
Removed because remoteWKClient is no longer optional